x/gamm/types/pool_asset.go

```diff
@@ -64,75 +75,28 @@ func SortPoolAssetsByDenom(assets []PoolAsset) {
 	})
 }
 
-// Validates a pool asset, to check if it has a valid weight.
-func (asset PoolAsset) ValidateWeight() error {
-	if asset.Weight.LTE(sdk.ZeroInt()) {
-		return fmt.Errorf("a token's weight in the pool must be greater than 0")
+func ValidateUserSpecifiedPoolAssets(assets []PoolAsset) error {
+	// The pool must be swapping between at least two assets
+	if len(assets) < 2 {
+		return ErrTooFewPoolAssets
 	}
 
-	// TODO: Choose a value that is too large for weights
-	// if asset.Weight >= (1 << 32) {
-	// 	return fmt.Errorf("a token's weight in the pool must be less than 2^32")
-	// }
-
-	return nil
-}
-
-// subPoolAssetWeights subtracts the weights of two different pool asset slices.
-// It assumes that both pool assets have the same token denominations,
-// with the denominations in the same order.
-// Returned weights can (and probably will have some) be negative.
-func subPoolAssetWeights(base []PoolAsset, other []PoolAsset) []PoolAsset {
-	weightDifference := make([]PoolAsset, len(base))
-	// TODO: Consider deleting these panics for performance
-	if len(base) != len(other) {
-		panic("subPoolAssetWeights called with invalid input, len(base) != len(other)")
-	}
-	for i, asset := range base {
-		if asset.Token.Denom != other[i].Token.Denom {
-			panic(fmt.Sprintf("subPoolAssetWeights called with invalid input, "+
-				"expected other's %vth asset to be %v, got %v",
-				i, asset.Token.Denom, other[i].Token.Denom))
-		}
-		curWeightDiff := asset.Weight.Sub(other[i].Weight)
-		weightDifference[i] = PoolAsset{Token: asset.Token, Weight: curWeightDiff}
+	// TODO: Add the limit of binding token to the pool params?
+	if len(assets) > 8 {
+		return sdkerrors.Wrapf(ErrTooManyPoolAssets, "%d", len(assets))
 	}
-	return weightDifference
-}
 
-// addPoolAssetWeights adds the weights of two different pool asset slices.
-// It assumes that both pool assets have the same token denominations,
-// with the denominations in the same order.
-// Returned weights can be negative.
-func addPoolAssetWeights(base []PoolAsset, other []PoolAsset) []PoolAsset {
-	weightSum := make([]PoolAsset, len(base))
-	// TODO: Consider deleting these panics for performance
-	if len(base) != len(other) {
-		panic("addPoolAssetWeights called with invalid input, len(base) != len(other)")
-	}
-	for i, asset := range base {
-		if asset.Token.Denom != other[i].Token.Denom {
-			panic(fmt.Sprintf("addPoolAssetWeights called with invalid input, "+
-				"expected other's %vth asset to be %v, got %v",
-				i, asset.Token.Denom, other[i].Token.Denom))
+	for _, asset := range assets {
+		err := ValidateUserSpecifiedWeight(asset.Weight)
+		if err != nil {
+			return err
 		}
-		curWeightSum := asset.Weight.Add(other[i].Weight)
-		weightSum[i] = PoolAsset{Token: asset.Token, Weight: curWeightSum}
-	}
-	return weightSum
-}
 
-// assumes 0 < d < 1
-func poolAssetsMulDec(base []PoolAsset, d sdk.Dec) []PoolAsset {
```

Comment:
Do we just no longer need these pool asset arithmetic methods?

Reply:
this was moved to `balancer_pool.go` bc the concept of pool weight moved to balancer package